File: CardMasterSkin/Elements/SERoundedRectangle.vb
Imports System.Drawing.Drawing2D
Imports CardMasterCard.Card
Imports CardMasterSkin.GraphicsElement

Namespace Skins

    Public Class SERoundedRectangle
        Inherits SkinElement

        Private m_cornerRadius As Integer

        Public Sub New(skin As Skin, width As Integer, height As Integer, cornerRadius As Integer)
            Me.New(skin, 0, 0, width, height, cornerRadius)
        End Sub

        Public Sub New(skin As Skin, x As Integer, y As Integer, width As Integer, height As Integer, cornerRadius As Integer)
            MyBase.New(skin, x, y, width, height)

            m_cornerRadius = cornerRadius

        End Sub

        Public Function GetCornerRadius() As Integer
            Return m_cornerRadius
        End Function

        Public Sub SetCornerRadius(cornerRadius As Integer)
            m_cornerRadius = cornerRadius
        End Sub

        Protected Overrides Function GetGraphicElements(card As Card) As List(Of GraphicElement)

            Dim graphicElementsList As New List(Of GraphicElement)()

            Dim path As New GraphicsPath()

            Dim w As Integer = Width
            Dim h As Integer = Height
            Dim b As Integer = GetCornerRadius()
            Dim r As Integer = 2 * b

            path.StartFigure()

            ' Coin haut/gauche
            path.AddArc(0, 0, r, r, 180, 90)
            ' Côté haut
            path.AddLine(b, 0, w - b, 0)
            ' Coin haut/droit
            path.AddArc(w - r, 0, r, r, 270, 90)
            ' Côté droit
            path.AddLine(w, b, w, h - b)
            ' Coin bas/droit
            path.AddArc(w - r, h - r, r, r, 0, 90)
            ' Côté bas
            path.AddLine(w - b, h, b, h)
            ' Coin bas/gauche
            path.AddArc(0, h - r, r, r, 90, 90)

            path.CloseFigure()

            graphicElementsList.Add(New PathElement(path, GetBackground))

            path = Nothing

            Return graphicElementsList

        End Function

    End Class

End Namespace

File: CardMasterImageBuilder/Skins/SkinElements/SERoundedRectangle.cs
using CardMasterCard.Card;
using CardMasterImageBuilder.GraphicsElements;
using CardMasterImageBuilder.Skins;
using System.Collections.Generic;
using System.Drawing.Drawing2D;
using System.IO;

//
//         270
//          |
//          |
// 180 -----+----- 0
//          |
//          |
//          0
//

namespace CardMasterImageBuilder.SkinElements
{
    public class SERoundedRectangle : SkinElement
    {
        public int CornerRadius { get; set; } = 0;

        public SERoundedRectangle(DirectoryInfo resourcesDirectory, int x, int y, int width, int height, string comments, int cornerRadius) : base(resourcesDirectory, x, y, width, height, comments)
        {
            this.CornerRadius = cornerRadius;
        }
        protected override List<GraphicElement> GetGraphicElements(JsonCard card)
        {
            var graphicElementsList = new List<GraphicElement>();
            var path = new GraphicsPath();

            int x = this.X;
            int y = this.Y;
            int w = Width;
            int h = Height;
            int b = this.CornerRadius;
            int r = 2 * b;

            // début de la forme
            path.StartFigure();

            // Coin haut/gauche
            path.AddArc(x, y, r, r, 180, 90);
            // Côté haut
            path.AddLine(x + b, y, w - b, y);
            // Coin haut/droit
            path.AddArc(x + w - r, y, r, r, 270, 90);
            // Côté droit
            path.AddLine(x + w, y + b, x + w, y + h - b);
            // Coin bas/droit
            path.AddArc(x + w - r, y + h - r, r, r, 0, 90);
            // Côté bas
            path.AddLine(x + w - b, y + h, x + b, y + h);
            // Coin bas/gauche
            path.AddArc(x, y + h - r, r, r, 90, 90);
            // Côté gauche
            // Généré par la fermeture de la forme

            // Fermeture de la forme
            path.CloseFigure();

            graphicElementsList.Add(new PathElement(path, GetBackground(path)));

            path = null;

            return graphicElementsList;

        }

    }
}
